### ISSUE TITLE
The term "Container" is now overloaded and underspecified

### ISSUE THREAD

**Issue by sandhawke** [solid/solid#37]
Some related problems:

Solid "containers" add many features to ldp:BasicContainers, most obviously having the URLs related to containment.   We've talked about proposing ldp:HierarchicalContainer as a flag for this.  There's also globbing, SPARQL, and probably several other features added.   Within the protocol we need names for these features, or collections of these features, so we can detect them.

In NoSQL databases like Mongo, I think these things are usually called Collections, not Containers.

The term "container" has recently become strongly attached to OS-level virtualization (LXC, vagrant, docker, sandstorm).  I've heard people confused by solid containers because they think we're talking about these.

Suggestion: in the documentation and toolkits, call them Folders.   Maybe in the HTTP headers, too, when we get there.

**Comment by dmitrizagidulin** [comment]
Agreed here. I like `Folders` more than `Containers`, but folders, too, is quite overloaded (in that, it suggests a filesystem-specific implementation). What if we called them `Collections`, straight up?

**Comment by sandhawke** [comment]
To my ear, Collections doesn't have the strictly-hierarchical connotation of Folders, and that strict hierarchy is the key point.

I don't get the filesystem connotation of Folders.  To me I just think of the icon.    If we called them Directories, which is also tempting, that would be a bit too close to the filesystem.   (Again, to my ear, and my reading of Wikipedia)

Slightly related, I'm trying to imagine if the root of a pod presented the public a display of the top level contents.   Kind of like a desktop, a file browser, and a nice website all at once.   Like you walk into a store, and you see the things in the store.    A radical idea, but potentially useful.   Trying to get rid of index.html.

**Comment by dmitrizagidulin** [comment]
Is the hierarchical nature of Containers a central point of them? If so, I didn't realize that. (And am ok with Folders).

**Comment by sandhawke** [comment]
There are a lot of details I don't remember offhand, but I remember hierarchy being at the heart of most of the disagreements between Tim's notion of how this should look and the LDP working group's notions.   At some point we need to enumerate all these, maybe soon.

**Comment by csarven** [comment]
If containers is to folders, is resources to papers? Sorry, a strong -1 from me for "folders".

I can't think of a decent replacement for Container... Would "Set" convey the same idea as a Container? Sets can be ordered or not and may contain other sets...

**Comment by dmitrizagidulin** [comment]
Why papers? You have folders on your local disk, on your OS, you dont call the files 'papers'.

**Comment by sandhawke** [comment]
@csarven Not papers, no, but... as it turns out... I do call them pages.    In general, I use the term "page" (or "webpage") instead of "resource", as do a lot of other people.

**Comment by csarven** [comment]
"folders", "papers", "pages" are all artefacts as we imitated the past. Can we move forward? :) This is no different than using icons like the pencil, or the floppy disk to convey certain operations which really don't have much to do with the medium in which they are in.

[Please excuse me, but Amy and I still have a "paper" to "write" :) ]

**Comment by sandhawke** [comment]
@csarven :-)    Yeah, we'll talk about this when that's done.

**Comment by sandhawke** [comment]
To give more evidence that we need to change away from "containers", here's email I just got, announcing a new meetup group:

*A group for people interested in talking about and hacking on Kubernetes, Google’s solution for scheduling and orchestrating containers at scale. We’re excited about microservices, containers, the distributions that run them and the solutions that deploy, manage, and extend them. Any skill level is welcome; we’re all new to Kubernetes and we want to create an open, welcoming environment for other Kubernauts....*

**Comment by bblfish** [comment]
Please check [LDP issue 50](https://www.w3.org/2012/ldp/track/issues/50) for further reasons for having what you wish to call `Folder`s, which I think is actually a good name (better than `iContainer` I suggested there).

**Comment by bblfish** [comment]
I'd suggest:

```Turtle
:Folder rdfs:subClassOf ldp:Container .
```

A `Folder` is just a restriction on an ldp:Container to make it work hierarchically and the way a folder on the file system works.

Are there useful `ldp:Container's that are not `:Folder`s . Yes: for example an `ldp:Container` that creates subdomains. That is actually why the restrictions for ldp:Containers were not made to be stricter.  I have a container that allows a POST to `http://example.com/` to create `http://joe.example.com/`.

**Comment by csarven** [comment]
@sandhawke Can you clarify the intention behind "folder"? Is it intended to include virtual folders? Or is the intention more like a directory in a filesystem? If it is more like a directory, I suggest sticking to well.. "directory", which is what we currently employ (although not part of the vocab) posix:Directory.

[Aside: every time I hear/read the term "folder", I associate it with the icon that looks like a folder IRL. maybe that's a good/bad thing... and perhaps just me]

**Comment by deiu** [comment]
+1 to posix:Directory (which we've used for a while). We need to check the semantics though to avoid having it mean a real directory on a file system.

**Comment by sandhawke** [comment]
There are perhaps two separate issues here, really.  

1.  What exactly are the properties of this thing?    Arguably, there might be more than one "thing" here, but let's try for one.

2.  What is a good name for this thing?  Arguably, there might be multiple names, such as in the HTTP headers, in the libraries, and in the user interfaces.  I think if we can find one that's acceptable in all these contexts that would be tremendously valuable.

I don't know all the details for #1 off the top of my head, but basically they're like ldp:Container with that addition of being hierarchically named.   That is:

- being "in" this thing is (mostly) equivalent to having a URL that adds one more path segment to the URL of this thing.   If you PUT with such a URL, where there's no member, you just added a member.  If you delete the thing, you delete all its members.   (I say "mostly" because I think acl and metadata resources don't follow this rule.)

- they're recursive.    If the thing is called $x/ and you PUT to $x/a/b/c the server should create $x/a/ and $x/a/b/ before creating $x/a/b/c.   Similarly, if you delete $x/ it MUST delete all resources with URLs starting $x/

This sounds a lot like a directory in a filesystem to those of us who type cd and mkdir -p and rm -rf a lot, but I note that more and more consumer facing software has decided to call these things folders instead of directories.

I'm not sure about posix:Directory issues.

Is there any observable difference between a "virtual folder" and a folder where a robot is helping add/remove things based on their properties?

**Comment by sandhawke** [comment]
@bblfish re ldp issue 50, I agree except given how other ldp issues sorted out, we'd now use <..> to refer to the container.

**Comment by dmitrizagidulin** [comment]
> What exactly are the properties of this thing?

The most important property of this thing is that it contains Resources :)
Which is why I think Collection is an appropriate term. (The hierarchical thing is just convenient namespacing -- incidental).

**Comment by bblfish** [comment]
@dmitrizagidulin  that's the definition of ldp:Collection. 
What we are looking here is for something with more specific properties than ldp:Collection. One does not exclude the other.

**Comment by dmitrizagidulin** [comment]
@bblfish - can you link me to the namespace/schema of `ldp:Collection`? I'm having trouble finding it. (Wasn't here https://www.w3.org/ns/ldp or here https://www.w3.org/TR/ldp/, etc)

**Comment by bblfish** [comment]
Sorry I meant `ldp:Container`. That's the super class of all containers.

**Comment by dmitrizagidulin** [comment]
No prob.

Given that we're trying to *rename* the term 'Container' (due to it having an overloaded meaning (Docker etc)), however, we can't really use that. 

(Collection, on the other hand, corresponds to what we're using these containers to implement).

**Comment by bblfish** [comment]
deleting.

**Comment by dmitrizagidulin** [comment]
What we're looking for is a *replacement* for the term 'Container'. (Since that's semantically over-loaded in most developers' minds (to mean Docker etc)). 

Hence - 'Collection'. Since that's what we're using these containers to implement -- collections of resources.

**Comment by sandhawke** [comment]
As I keep saying, I want to solve several different problems at once, because Folder happens to do that.   The two big problems are that container suddenly now means something else in our industry and that our Things are strictly hierarchical, which is the main way they are different from everyone else's Things.

**Comment by bblfish** [comment]
I'd say we have consensus on needing to create a new subclass of `ldp:Container` that gives us hierarchical containers that fit the URL spec (as specified [ldp issue 50](https://www.w3.org/2012/ldp/track/issues/50)). What to call it is still open.
Do we agree? If so we can then next debate, vote for the name.

**Comment by cristianvasquez** [comment]
Hi,

I'm looking for a definition of what is a Container, perhaps is useful for my use case.

In my use case I'm trying to figure out how to define groups of resources, or resource-spaces.
The data itself can reside in a file system, a triple-store, quadstore etc.

When the implementation is file-based I can use the normal GRUB patterns to define the groups of resources

/people/alice/music/drums.ttl (resource-space: drum music of alice)
/people/*/music/*  (resource-space: all the music of all people)

When the implementation is a triplestore, I can define this space as a SPARQL query

When the implementation is a quadstore, I can define this space as sets of named graphs

There is something like this in the specs?
